### ISSUE TITLE
Base.hasfastin(Iterators.Pairs) needs to check how it is storing the keys

### ISSUE THREAD

**Issue by oxinabox** [JuliaLang/julia#34672]
Similar to https://github.com/JuliaCollections/OrderedCollections.jl/issues/33

We see that:
```
julia> x = pairs((a=1,b=2))
pairs(::NamedTuple) with 2 entries:
  :a => 1
  :b => 2

julia> Base.hasfastin(typeof(x))
true

julia> Base.hasfastin(typeof(x.itr))
false
```

`hasfastin`  if the checking of `haskey` is expected logorithmic or better.
but for for many uses of pairs, e.g. when applied to a `NamedTuple` it is linear.

But we are falling back to the default for `AbstractDict` which is `true`.

https://github.com/JuliaLang/julia/blob/aa36eefa17e68dd3781b668bc40328583b161924/base/abstractset.jl#L284-L295

**Comment by rfourquet** [comment]
I can also think of `Base.ImmutableDict` that has only linear lookup. On the other hand, my impression is that `Base.ImmutableDict` is typically used only with few elements, and `NamedTuple` too, which would make me think that "linear" in this case is still "fast" (and maybe for `LittleDict` too, AFAIR, it's supposed to be used mostly with few elements). 

It would be great to have one (or more) benchmarks of the implication of changing this `hasfastin` trait on real use-cases.

**Comment by oxinabox** [comment]
I think the idea is on does something like:

```
if length(x) < 32 || hasfastin(typeof(x))
    dothing(keys(x))
else
    dothing(Set(keys(x)))
end
```

**Comment by rfourquet** [comment]
> I think the idea is on does something like:

Precisely, so for small collections it might still be better to not create a `Set` even when they have only linear lookup.

**Comment by oxinabox** [comment]
example of use:
https://github.com/JuliaLang/julia/blob/3aeb3ec9ee8d37b16187f4b3ae52389fedf253a9/base/abstractset.jl#L274-L280
